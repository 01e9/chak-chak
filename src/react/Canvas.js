import React from 'react'
import PropTypes from 'prop-types'
import { canvasDrawImage } from "@/utility/canvas";
import { withStyles } from '@material-ui/core'

class Canvas_ extends React.Component {
    static propTypes = {
        classes: PropTypes.object.isRequired,
        image: PropTypes.instanceOf(Image).isRequired
    }

    canvasRef = React.createRef()

    componentDidMount() {
        canvasDrawImage(this.canvasRef.current, this.props.image);
    }

    render() {
        const { classes } = this.props;
        return (
            <canvas ref={this.canvasRef} className={classes.root} />
        );
    }
}

const styles = {
    root: {
        display: 'block'
    }
}

const Canvas = withStyles(styles)(Canvas_);

export default Canvas;
