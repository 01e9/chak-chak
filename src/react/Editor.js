import React from 'react'
import PropTypes from 'prop-types'
import { withStyles } from '@material-ui/core'

class Editor_ extends React.Component {
    static propTypes = {
        classes: PropTypes.object.isRequired,
        Canvas: PropTypes.func.isRequired,
        Toolbar: PropTypes.func.isRequired,
        ActiveTool: PropTypes.func.isRequired
    }

    render() {
        const { Canvas, Toolbar, ActiveTool, classes } = this.props;

        return (
            <React.Fragment>
                <div className={classes.canvasWrap}>
                    <Canvas/>
                    <ActiveTool/>
                </div>
                <Toolbar/>
            </React.Fragment>
        )
    }
}

const styles = {
    canvasWrap: {
        position: 'relative'
    }
}

const Editor = withStyles(styles)(Editor_);

export default Editor;
