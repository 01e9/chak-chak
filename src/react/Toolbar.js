import React from 'react'
import PropTypes from 'prop-types'
import { withStyles } from '@material-ui/core'
import Paper from '@material-ui/core/Paper';
import Tabs from '@material-ui/core/Tabs';
import Tab from '@material-ui/core/Tab';
import {
    DragIndicator as DragIndicatorIcon,
    Brush as BrushIcon,
    Crop as CropIcon,
    CloudUpload as CloudUploadIcon
} from '@material-ui/icons';

class Toolbar_ extends React.Component {
    static propTypes = {
        classes: PropTypes.object.isRequired
    }

    render() {
        const { classes } = this.props;
        const tabProps = {
            classes: {root: classes.tab}
        }

        return (
            <Paper className={classes.paper}>
                <Tabs>
                    <Tab icon={<DragIndicatorIcon/>} {...tabProps} />
                    <Tab icon={<BrushIcon/>} {...tabProps} />
                    <Tab icon={<CropIcon/>} {...tabProps} />
                    <Tab icon={<CloudUploadIcon/>} {...tabProps} />
                </Tabs>
            </Paper>
        )
    }
}

const styles = {
    paper: {
        margin: '100px',
        display: 'inline-block'
    },
    tab: {
        minWidth: '50px'
    }
}

const Toolbar = withStyles(styles)(Toolbar_)

export default Toolbar
