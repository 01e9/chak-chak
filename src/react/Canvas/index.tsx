import * as React from 'react'
import { canvasDrawImage } from "@/utility/canvas";
import * as styles from "./styles.scss";

export interface ICanvasProps {
    image?: HTMLImageElement;
    aspectRatio: number;
}

export default class Canvas extends React.Component<ICanvasProps> {
    canvasRef = React.createRef<HTMLCanvasElement>()

    redrawImage() {
        if (this.canvasRef.current && this.props.image) {
            canvasDrawImage(this.canvasRef.current, this.props.image, this.props.aspectRatio);
        }
    }

    componentDidMount() {
        this.redrawImage();
    }

    componentDidUpdate() {
        this.redrawImage();
    }

    render() {
        return (
            <canvas ref={this.canvasRef} className={styles.root} />
        );
    }
}
