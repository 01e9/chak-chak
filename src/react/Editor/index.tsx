import * as React from 'react'
import * as styles from "./styles.scss";

export interface IEditorProps {
    Canvas: React.JSXElementConstructor<{}>,
    Toolbar: React.JSXElementConstructor<{}>,
    ActiveTool: React.JSXElementConstructor<{}>
}

class Editor extends React.Component<IEditorProps> {
    render() {
        const { Canvas, Toolbar, ActiveTool } = this.props;

        return (
            <React.Fragment>
                <div className={styles.canvasWrap}>
                    <Canvas/>
                    <ActiveTool/>
                </div>
                <Toolbar/>
            </React.Fragment>
        )
    }
}

export default Editor;
