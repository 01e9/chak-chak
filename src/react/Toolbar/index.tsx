import * as React from 'react'
import Paper from '@material-ui/core/Paper';
import Button from '@material-ui/core/Button';
import { DragIndicator as DragIndicatorIcon } from '@material-ui/icons';
import * as styles from "./styles.scss";

export interface IToolbarProps {
    left: number;
    top: number;
    onPositionUpdate: (left: number, top: number) => void;
    tools: Array<{key: string; Component: React.JSXElementConstructor<{}>}>;
}

interface IState {
    hasCapture: boolean;
    left: number;
    top: number;
    leftShift: number;
    topShift: number;
}

class Toolbar extends React.Component<IToolbarProps, IState> {
    state: IState = {
        hasCapture: false,
        left: 0,
        top: 0,
        leftShift: 0,
        topShift: 0
    }

    static getDerivedStateFromProps(props: IToolbarProps, state: IState) {
        if (state.hasCapture) {
            return null;
        } else {
            return {
                left: props.left,
                top: props.top
            }
        }
    }

    getPointerPosition(event: React.PointerEvent) {
        return {
            left: event.pageX,
            top: event.pageY
        }
    }

    onDown = (event: React.PointerEvent) => {
        // @ts-ignore setPointerCapture is not defined
        event.target.setPointerCapture(event.pointerId);
    }

    onGotCapture = (event: React.PointerEvent) => {
        const pointerPosition = this.getPointerPosition(event);

        this.setState(({left, top}) => ({
            hasCapture: true,
            leftShift: left - pointerPosition.left,
            topShift: top - pointerPosition.top
        }));
    }

    onMove = (event: React.PointerEvent) => {
        event.stopPropagation();
        const pointerPosition = this.getPointerPosition(event);

        this.setState(({leftShift, topShift}) => ({
            left: pointerPosition.left + leftShift,
            top: pointerPosition.top + topShift
        }));
    }

    onLostCapture = () => {
        this.setState({
            hasCapture: false,
            leftShift: 0,
            topShift: 0
        });

        this.props.onPositionUpdate(
            Math.max(0, this.state.left),
            Math.max(0, this.state.top)
        );
    }

    render() {
        const
            { tools } = this.props,
            { top, left, hasCapture } = this.state;

        return (
            <Paper className={styles.root} style={{top: top + 'px', left: left + 'px'}}>
                <Button
                    className={styles.drag + ' drag-button'}
                    disableRipple

                    onPointerDown={this.onDown}
                    onGotPointerCapture={this.onGotCapture}
                    onPointerMove={hasCapture ? this.onMove : undefined}
                    onLostPointerCapture={this.onLostCapture}
                >
                    <DragIndicatorIcon />
                </Button>
                {tools.map(({key, Component}) => <Component key={key} />)}
            </Paper>
        )
    }
}

export default Toolbar;
