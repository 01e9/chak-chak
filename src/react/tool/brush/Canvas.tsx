import * as React from "react";

interface IToolBrushCanvasProps {
    width: number;
    height: number;
}
