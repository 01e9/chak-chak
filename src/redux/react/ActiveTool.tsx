import * as React from "react";
import { TOOL_NAME_BRUSH } from "@/constants/tools";
import { connect } from "react-redux";
import { IState } from "@/redux/reducers";

interface IInternalProps {
    activeTool: string;
}

const mapStateToProps = ({ activeTool }: IState) => ({ activeTool });

const render = ({ activeTool }: IInternalProps) => {
    switch (activeTool) {
        case TOOL_NAME_BRUSH: return <div>BRUSH</div>;
        default: return null;
    }
};

export default connect(mapStateToProps)(render);
