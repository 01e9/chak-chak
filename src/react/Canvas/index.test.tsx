import * as React from 'react';
import { createShallow } from '@material-ui/core/test-utils';
import Canvas, { ICanvasProps } from "@/react/Canvas";

const requiredProps: ICanvasProps = {
    image: new Image(100, 100),
    aspectRatio: 1,
}

describe('snapshot', () => {
    it('required props', () => {
        const wrapper = createShallow()(<Canvas {...requiredProps} />);
        expect(wrapper).toMatchSnapshot()
    })
    it('changed image', async () => {
        const wrapper = createShallow()(<Canvas {...requiredProps} />);
        expect(wrapper).toMatchSnapshot("before")
        await new Promise(resolve => wrapper.setProps({image: new Image(101, 102)}, resolve))
        expect(wrapper).toMatchSnapshot("after")
    })
})
