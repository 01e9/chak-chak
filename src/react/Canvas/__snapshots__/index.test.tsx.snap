// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot changed image: after 1`] = `<canvas />`;

exports[`snapshot changed image: before 1`] = `<canvas />`;

exports[`snapshot required props 1`] = `<canvas />`;
